### 📘 ¿Qué es un Modelo de Clasificación y un Modelo de Regresión?

En Machine Learning, un **modelo predictivo** se entrena con datos históricos para hacer predicciones sobre datos nuevos. Dependiendo del tipo de predicción que queremos hacer, usamos **dos tipos principales de modelos**:

---

#### 🔢 Modelo de Regresión

Un modelo de **regresión** se utiliza cuando queremos predecir un **valor numérico continuo**.

📌 **Ejemplos:**
- ¿Cuántos turistas llegarán a Cartagena el próximo mes?
- ¿Cuál será el costo promedio del alojamiento?

🎯 **Objetivo:** Predecir cantidades (números).

---

#### 🏷️ Modelo de Clasificación

Un modelo de **clasificación** se utiliza cuando queremos predecir una **categoría o clase**.

📌 **Ejemplos:**
- ¿La ocupación hotelera será *Alta* o *Baja*?
- ¿Habrá *lluvia* o *no lluvia*?

🎯 **Objetivo:** Predecir etiquetas (clases).

---

### 🤔 ¿Cómo sé cuál modelo usar?

| Pregunta clave                          | Si la respuesta es...        | Usa...          |
|----------------------------------------|-------------------------------|------------------|
| ¿Lo que quiero predecir es un número?  | ✅ Sí                          | 🔢 Regresión     |
| ¿Lo que quiero predecir es una clase?  | ✅ Sí (*Alta/Baja*, *Sí/No*)   | 🏷️ Clasificación |

---

💡 En este taller vamos a usar **ambos tipos de modelos**:

- 🔢 Un modelo de **regresión** para predecir el **número de turistas** que visitan Cartagena.
- 🏷️ Un modelo de **clasificación** para predecir si la **ocupación hotelera** será *Alta* o *Baja*.


### 🎯 ¿Qué son las MÉTRICAS en Machine Learning?

Las **métricas** son herramientas que nos permiten **evaluar el rendimiento de un modelo**.  
Nos ayudan a saber si el modelo hace buenas predicciones o si se está equivocando, y **cuánto se equivoca**.

Las métricas que se usan dependen del tipo de modelo que estamos usando:  
🔢 **Regresión** o 🏷️ **Clasificación**.

---

## 📈 MÉTRICAS PARA REGRESIÓN

Se usan cuando el modelo predice **valores numéricos continuos**, como el número de turistas.

| Métrica        | ¿Qué mide?                                                           | ¿Qué significa un valor bajo?         |
|----------------|----------------------------------------------------------------------|---------------------------------------|
| **MAE**        | *Error Absoluto Medio*: promedio de las diferencias absolutas entre las predicciones y los valores reales. | El modelo tiene errores pequeños.     |
| **MSE**        | *Error Cuadrático Medio*: igual que MAE pero eleva los errores al cuadrado (penaliza errores grandes).     | El modelo comete pocos errores graves.|
| **RMSE**       | *Raíz del MSE*: vuelve a la escala original de los datos, fácil de interpretar.                             | El modelo predice con buena precisión.|

📌 **Ejemplo**:  
Si el RMSE = 1500, significa que el modelo se equivoca en promedio **en 1.500 turistas**.

---

## 🏷️ MÉTRICAS PARA CLASIFICACIÓN

Se usan cuando el modelo predice **categorías**, como *Alta* o *Baja* ocupación hotelera.

| Métrica         | ¿Qué mide?                                                                 | Interpretación                          |
|------------------|---------------------------------------------------------------------------|------------------------------------------|
| **Accuracy**     | Porcentaje de predicciones correctas.                                     | A mayor valor, mejor modelo.             |
| **Precision**    | De todas las veces que el modelo dijo “Alta”, ¿cuántas fueron ciertas?    | Evita falsos positivos.                  |
| **Recall**       | De todas las veces que realmente fue “Alta”, ¿cuántas veces acertó el modelo? | Evita falsos negativos.              |
| **F1-score**     | Promedio entre Precision y Recall.                                        | Equilibrio cuando hay clases desbalanceadas.|

📌 **Ejemplo**:  
Si el modelo dijo "Alta" 10 veces, pero solo 7 eran ciertas:  
- **Precision** = 7/10 = 0.70  
Si realmente había 14 casos de "Alta" y el modelo predijo 7:  
- **Recall** = 7/14 = 0.50

---

### 🧠 ¿Cuál métrica usar?

| Si te importa...                        | Usa esta métrica               |
|----------------------------------------|-------------------------------|
| Que el modelo acierte en general       | Accuracy                      |
| Que los “positivos” sean correctos     | Precision                     |
| Que no se te escapen los “positivos”   | Recall                        |
| Un equilibrio entre ambos              | F1-score                      |


In [110]:
import pandas as pd
import numpy as np

# Cargar archivo
df = pd.read_csv('turistas_cartagena_data.csv')

# Guarda una copia del dataset original antes de comenzar la limpieza.
df_original = df.copy()

# Vista general
print(df.info())
df.head(20)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Fecha                                      495 non-null    object 
 1   Número de Visitantes                       485 non-null    float64
 2   Temperatura Media (°C)                     495 non-null    float64
 3   Precipitación (mm)                         495 non-null    float64
 4   Eventos Especiales                         495 non-null    float64
 5   Costo Promedio de Alojamiento (COP)        495 non-null    float64
 6   Tasa de Cambio (USD/COP)                   495 non-null    float64
 7   Promociones de Turismo                     495 non-null    float64
 8   Cantidad de Vuelos y Cruceros Disponibles  495 non-null    float64
 9   Ocupación Hotelera (%)                     495 non-null    float64
dtypes: float64(9), object(1)
m

Fecha  Número de Visitantes  Temperatura Media (°C)  Precipitación (mm)  \
0   1982-12              126958.0                    27.6               195.4   
1   1983-01              136932.0                    27.6               248.9   
2   1983-02              108694.0                    30.7                61.9   
3   1983-03              124879.0                    31.6                 3.3   
4   1983-04              115268.0                    31.9                41.1   
5   1983-05               59886.0                    30.3               270.0   
6       NaN                   NaN                     NaN                 NaN   
7       NaN                   NaN                     NaN                 NaN   
8       NaN                   NaN                     NaN                 NaN   
9       NaN                   NaN                     NaN                 NaN   
10      NaN                   NaN                     NaN                 NaN   
11  1983-11               46090.0                    31.3               274.3   
12  1983-12                   NaN                    25.8               125.6   
13  1983-12               69820.0                    28.4               114.9   
14  1984-01                5769.0                    25.1               155.7   
15  1984-02               64735.0                    28.3               -50.0   
16  1984-03               69925.0                    25.4                49.9   
17  1984-04               10311.0                    25.8               221.4   
18  1984-05              108355.0                    25.8                24.8   
19  1984-06               90305.0                    29.5               180.9   

    Eventos Especiales  Costo Promedio de Alojamiento (COP)  \
0                  0.0                             252889.0   
1                  0.0                             268933.0   
2                  1.0                             113428.0   
3                  0.0                             275006.0   
4                  0.0                             346627.0   
5                  1.0                             345439.0   
6                  NaN                                  NaN   
7                  NaN                                  NaN   
8                  NaN                                  NaN   
9                  NaN                                  NaN   
10                 NaN                                  NaN   
11                 0.0                             237328.0   
12                 0.0                             355158.0   
13                 1.0                             176256.0   
14                 0.0                             461278.0   
15                 0.0                             232723.0   
16                 0.0                             281360.0   
17                 0.0                             446874.0   
18                 0.0                             389530.0   
19                 0.0                             342634.0   

    Tasa de Cambio (USD/COP)  Promociones de Turismo  \
0                    3076.77                     0.0   
1                    3231.50                     1.0   
2                    3478.34                     0.0   
3                    3925.80                     0.0   
4                    3400.98                     0.0   
5                    3615.39                     0.0   
6                        NaN                     NaN   
7                        NaN                     NaN   
8                        NaN                     NaN   
9                        NaN                     NaN   
10                       NaN                     NaN   
11                   3208.74                     1.0   
12                   2568.47                     1.0   
13                   2710.04                     1.0   
14                   2915.22                     0.0   
15                   3957.30                     1.0   
16                   2997.02

**🧽 Paso 2: Limpieza de Datos**

In [111]:
#imprimir registros con valores nulos  a lo ancho
print(df[df.isnull().any(axis=1)])


# Ver valores nulos
print(df.isnull().sum())

# Eliminar filas completamente vacías
df.dropna(how='all', inplace=True)


# Rellenar 'Número de Visitantes' con la media
df['Número de Visitantes'].fillna(df['Número de Visitantes'].mean(), inplace=True)

#imprimir


       Fecha  Número de Visitantes  Temperatura Media (°C)  \
6        NaN                   NaN                     NaN   
7        NaN                   NaN                     NaN   
8        NaN                   NaN                     NaN   
9        NaN                   NaN                     NaN   
10       NaN                   NaN                     NaN   
12   1983-12                   NaN                    25.8   
52   1987-03                   NaN                    28.2   
57   1987-08                   NaN                    31.9   
71   1988-10                   NaN                    31.1   
114  1992-04                   NaN                    25.9   
125  1993-03                   NaN                    28.1   
140  1994-06                   NaN                    28.4   
327  2009-10                   NaN                    30.7   
377  2013-11                   NaN                    27.3   
379  2014-01                   NaN                    25.0   

     Pr

2.2 Eliminar duplicados

In [112]:
# Identificar filas duplicadas
duplicados = df[df.duplicated()]


#imprimir data set sin datos duplicados
print(duplicados)

#eliminar duplicados
df.drop_duplicates(inplace=True)


Empty DataFrame
Columns: [Fecha, Número de Visitantes, Temperatura Media (°C), Precipitación (mm), Eventos Especiales, Costo Promedio de Alojamiento (COP), Tasa de Cambio (USD/COP), Promociones de Turismo, Cantidad de Vuelos y Cruceros Disponibles, Ocupación Hotelera (%)]
Index: []


2.3 Corregir valores inconsistentes

En este paso, se deben buscar valores que estén fuera de los rangos aceptables. Por ejemplo, una precipitación negativa no tiene sentido, así que hay que corregir estos valores:

In [113]:
# Identificar valores inconsistentes (precipitación negativa)
df_inconsistentes = df[df['Precipitación (mm)'] < 0]

# Corregir valores inconsistentes (estableciendo a 0 o la media, por ejemplo)
df.loc[df['Precipitación (mm)'] < 0, 'Precipitación (mm)'] = df['Precipitación (mm)'].mean()


2.4 Normalización (para modelos)
*  La normalización es importante para que los valores de las distintas columnas estén en rangos comparables, especialmente cuando se utilizan modelos de Machine Learning que se ven afectados por escalas de variables diferentes (por ejemplo, redes neuronales o métodos de distancia).

Dos técnicas comunes de normalización son:
* Min-Max Scaling: Escala los datos a un rango entre 0 y 1.
* Estandarización: Centra los datos en la media con una desviación estándar de 1.

In [114]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
columnas_a_normalizar = [
    'Temperatura Media (°C)', 'Precipitación (mm)',
    'Costo Promedio de Alojamiento (COP)', 'Tasa de Cambio (USD/COP)',
    'Cantidad de Vuelos y Cruceros Disponibles'
]
df[columnas_a_normalizar] = scaler.fit_transform(df[columnas_a_normalizar])
#print(df)


**🔢 Paso 3: Modelo de Regresión (Predecir visitantes)**  

3.1 Separar variables

### 📊 Principales Modelos de Regresión

| **Modelo de Regresión**           | **¿Cuándo usarlo?**                                                                 |
|----------------------------------|-------------------------------------------------------------------------------------|
| 🔹 **Regresión Lineal**          | Cuando la relación entre las variables es lineal (tendencia recta).                |
| 🔹 **Regresión Polinómica**      | Cuando la relación tiene forma de curva o cambia en distintos tramos.              |
| 🔹 **Regresión Ridge/Lasso**     | Cuando hay muchas variables y se desea evitar sobreajuste (regularización).        |
| 🔹 **Regresión con Árboles**     | Cuando la relación entre variables es no lineal y se requieren decisiones por rangos. |
| 🔹 **Random Forest Regressor**   | Cuando hay muchas variables y relaciones complejas; mejora precisión.              |
| 🔹 **Redes Neuronales**          | Cuando se trabaja con grandes volúmenes de datos y relaciones altamente complejas.  |

---

💡 En este taller vamos a usar el modelo de **Regresión Lineal**, ideal para comenzar por su simplicidad y facilidad de interpretación.



In [115]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

X = df[columnas_a_normalizar]
y = df['Número de Visitantes']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


3.2 Entrenar modelo

In [116]:


# Ahora puedes entrenar el modelo
modelo_regresion = LinearRegression()
modelo_regresion.fit(X_train, y_train)

LinearRegression()

3.3 Métricas básicas (explicadas)

## 📈 MÉTRICAS PARA REGRESIÓN

Se usan cuando el modelo predice **valores numéricos continuos**, como el número de turistas.

| Métrica        | ¿Qué mide?                                                           | ¿Qué significa un valor bajo?         |
|----------------|----------------------------------------------------------------------|---------------------------------------|
| **MAE**        | *Error Absoluto Medio*: promedio de las diferencias absolutas entre las predicciones y los valores reales. | El modelo tiene errores pequeños.     |
| **MSE**        | *Error Cuadrático Medio*: igual que MAE pero eleva los errores al cuadrado (penaliza errores grandes).     | El modelo comete pocos errores graves.|
| **RMSE**       | *Raíz del MSE*: vuelve a la escala original de los datos, fácil de interpretar.                             | El modelo predice con buena precisión.|

📌 **Ejemplo**:  
Si el RMSE = 1500, significa que el modelo se equivoca en promedio **en 1.500 turistas**.

In [117]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

y_pred = modelo_regresion.predict(X_test)

# Explicación para estudiantes:
# MAE: diferencia promedio entre valores reales y predichos.
# MSE: error cuadrático medio (penaliza más los errores grandes).
# RMSE: raíz del MSE, tiene la misma escala que los valores reales.

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

print(f"MAE: {mae:.2f}")
print(f"MSE: {mse:.2f}")
print(f"RMSE: {rmse:.2f}")



MAE: 37278.66
MSE: 1776346926.16
RMSE: 42146.73


In [118]:
# Calcular el promedio de los valores reales
promedio_reales = y_test.mean()

# Comparar el MAE con ese promedio
porcentaje_error_mae = (mae / promedio_reales) * 100


print(f"MAE representa un error del {porcentaje_error_mae:.2f}% respecto al promedio real de turistas.")

# Comparar el RMSE  con ese promedio
porcentaje_error_mae = (rmse / promedio_reales) * 100
print(f"RMSE representa un error del {porcentaje_error_mae:.2f}% respecto al promedio real de turistas.")


MAE representa un error del 48.10% respecto al promedio real de turistas.
RMSE representa un error del 54.38% respecto al promedio real de turistas.


### 📊 Lectura de las métricas aplicadas al modelo de regresión

Este análisis se basa en un modelo que predice **valores numéricos continuos**, específicamente el **número de turistas**.

---

#### 🧠 ¿Qué se midió?

Se aplicaron tres métricas para evaluar el rendimiento del modelo:

- **MAE (Error Absoluto Medio)**  
- **MSE (Error Cuadrático Medio)**  
- **RMSE (Raíz del Error Cuadrático Medio)**

---

#### 📌 ¿Qué significan los valores obtenidos?

- **MAE: 37,278.66**  
  → En promedio, el modelo se equivoca por **37 mil turistas**.  
  ✏️ *Este valor indica un error alto si se necesita precisión.*

- **MSE: 1,776,346,926.16**  
  → Penaliza más los errores grandes.  
  ✏️ *Este número es elevado y está en escala cuadrática, lo que lo hace difícil de interpretar directamente.*

- **RMSE: 42,146.73**  
  → Devuelve el error a la misma escala que los datos originales.  
  → En promedio, el modelo se equivoca por **42 mil turistas**.  
  ✏️ *Este valor es más fácil de entender y refleja que el modelo aún no es preciso.*

---

#### 🎯 Conclusión:

- Estas métricas permiten **cuantificar el error del modelo**.
- Un **RMSE mayor a 30 mil** puede ser inaceptable dependiendo del contexto.
- El modelo debería **mejorarse** para ofrecer predicciones más cercanas a la realidad.
- El **RMSE** es especialmente útil porque **está en la misma escala** que el número de turistas.

---

> 📌 **Nota**: Si el total estimado de turistas es, por ejemplo, 100.000, un error de 42.000 representa una desviación significativa que puede impactar la toma de decisiones.



✅ Bloque para probar el modelo con nuevos datos ingresados por el usuario

In [119]:
# Supongamos que ya entrenaste el modelo con estas variables:
# ['Temperatura Media (°C)', 'Precipitación (mm)',
#  'Costo Promedio de Alojamiento (COP)', 'Tasa de Cambio (USD/COP)',
#  'Cantidad de Vuelos y Cruceros Disponibles']

# Paso 1: Ingresar nuevos datos manualmente
nuevos_datos = {
    'Temperatura Media (°C)': [30],
    'Precipitación (mm)': [5],
    'Costo Promedio de Alojamiento (COP)': [200000],
    'Tasa de Cambio (USD/COP)': [4000],
    'Cantidad de Vuelos y Cruceros Disponibles': [50]
}

nuevo_df = pd.DataFrame(nuevos_datos)

# Paso 2: Normalizar los nuevos datos con el mismo scaler usado antes
nuevo_df_normalizado = scaler.transform(nuevo_df)

# Paso 3: Hacer la predicción
prediccion = modelo_regresion.predict(nuevo_df_normalizado)

# Paso 4: Mostrar el resultado
print(f"🔮 Número estimado de visitantes: {int(prediccion[0]):,} turistas")


🔮 Número estimado de visitantes: 85,843 turistas


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


**🏷️ Paso 4: Modelo de Clasificación (Alta/Baja Ocupación Hotelera)**

4.1 Crear variable categórica

In [120]:
# Umbral: 70% o más se considera "Alta"
df['Ocupación Categoría'] = df['Ocupación Hotelera (%)'].apply(lambda x: 'Alta' if x >= 70 else 'Baja')


4.2 Entrenar modelo

### ✅ ¿Qué modelo de clasificación debo emplear?

🎯 **1. ¿Qué tipo de problema de clasificación tengo?**

| **Situación**                                                                 | **Modelo recomendado**                      |
|-------------------------------------------------------------------------------|---------------------------------------------|
| Quiero clasificar en **dos categorías** (Ej: *Alta* o *Baja*)                | 🔹 Regresión Logística                      |
| Las relaciones entre variables no son lineales o hay muchas reglas complejas | 🔹 Árbol de Decisión                        |
| Tengo muchas variables y quiero mejorar la precisión                         | 🔹 Random Forest Classifier                 |
| Quiero un modelo rápido y eficiente para datos lineales                      | 🔹 Support Vector Machine (SVM)             |
| Tengo muchos datos y quiero un modelo que aprenda patrones complejos         | 🔹 Redes Neuronales                         |

---

💡 En este ejercicio trabajaremos con **Random Forest Classifier**, ideal para problemas de clasificación donde hay varias variables y relaciones complejas.


In [121]:
from sklearn.ensemble import RandomForestClassifier

X_clas = df[columnas_a_normalizar]
y_clas = df['Ocupación Categoría']

Xc_train, Xc_test, yc_train, yc_test = train_test_split(X_clas, y_clas, test_size=0.2, random_state=42)

modelo_clasificacion = RandomForestClassifier()
modelo_clasificacion.fit(Xc_train, yc_train)


RandomForestClassifier()

4.3 Métricas para clasificación (explicadas)

In [122]:
from sklearn.metrics import accuracy_score, classification_report

yc_pred = modelo_clasificacion.predict(Xc_test)

# Accuracy: porcentaje de predicciones correctas.
# Precision: cuántos positivos predichos realmente lo son.
# Recall: cuántos positivos reales se detectaron.
# F1-score: balance entre precision y recall.

print("Accuracy:", accuracy_score(yc_test, yc_pred))
print("\nReporte de Clasificación:")
print(classification_report(yc_test, yc_pred))


Accuracy: 0.5454545454545454

Reporte de Clasificación:
              precision    recall  f1-score   support

        Alta       0.57      0.75      0.65        55
        Baja       0.48      0.30      0.37        44

    accuracy                           0.55        99
   macro avg       0.53      0.52      0.51        99
weighted avg       0.53      0.55      0.52        99



### 📊 ¿Cómo interpretamos las métricas del modelo de clasificación?

Después de entrenar un modelo de clasificación como `RandomForestClassifier`, es importante **evaluar qué tan bien predice las categorías**. En este caso, estamos prediciendo si la **ocupación hotelera** será **Alta** o **Baja**.

---

### ✅ 1. Accuracy: **0.57**

- **¿Qué significa?** El modelo acertó en el **57%** de los casos.
- **¿Es bueno?** Un valor de 0.57 es bajo. Significa que el modelo se equivoca frecuentemente.
  
> 📉 *Necesitamos mejorarlo, tal vez con más datos, otro modelo o ajustando los parámetros.*

---

### 📋 2. Reporte por Clase

| Clase | Precision | Recall | F1-score | Cantidad de casos (support) |
|-------|-----------|--------|----------|------------------------------|
| **Alta** | 0.62 | 0.68 | 0.65 | 59 |
| **Baja** | 0.47 | 0.41 | 0.44 | 41 |

#### 🟩 Precision:
- **¿Qué mide?** Cuántas veces que el modelo dijo una clase, acertó.
- **Ejemplo Alta (0.62):** Cuando el modelo predijo “Alta”, acertó el **62%** de las veces.

#### 🟦 Recall:
- **¿Qué mide?** Cuántos casos reales de una clase el modelo logró detectar.
- **Ejemplo Alta (0.68):** De todos los casos reales de “Alta”, el modelo detectó el **68%**.

#### 🟣 F1-score:
- **¿Qué mide?** Un promedio entre Precision y Recall.
- **Ejemplo Alta (0.65):** Buen balance entre aciertos y detección.

> ⚠️ El modelo **funciona mejor para la clase "Alta"** que para "Baja". Detecta y acierta más en esos casos.

---

### 🔢 3. Promedios generales

| Tipo de promedio | Precision | Recall | F1-score |
|------------------|-----------|--------|----------|
| **Macro avg**     | 0.55      | 0.55   | 0.55     |
| **Weighted avg**  | 0.56      | 0.57   | 0.56     |

- **Macro avg:** Promedia cada clase por igual.
- **Weighted avg:** Toma en cuenta cuántos ejemplos hay de cada clase.

---

### 🎯 Conclusión para este modelo:

- Tiene **mayor precisión para predecir “Alta”** ocupación.
- Tiene **dificultades con la clase “Baja”**, pues falla más.
- Necesitamos mejorar el modelo si queremos que sea más confiable en ambas clases.

💡 *Posibles mejoras: ajustar parámetros, probar otros modelos o balancear las clases.*


Bloque para probar el modelo de clasificación con nuevos datos ingresados por el usuario

In [123]:
# Paso 1: Ingresar nuevos datos manualmente
nuevos_datos_clas = {
    'Temperatura Media (°C)': [30],
    'Precipitación (mm)': [10],
    'Costo Promedio de Alojamiento (COP)': [180000],
    'Tasa de Cambio (USD/COP)': [4100],
    'Cantidad de Vuelos y Cruceros Disponibles': [60]
}

nuevo_df_clas = pd.DataFrame(nuevos_datos_clas)

# Paso 2: Normalizar los nuevos datos con el mismo scaler
nuevo_df_clas_normalizado = scaler.transform(nuevo_df_clas)

# Paso 3: Hacer la predicción con el modelo Random Forest
prediccion_clas = modelo_clasificacion.predict(nuevo_df_clas_normalizado)

# Paso 4: Mostrar el resultado
print(f"🏨 Ocupación hotelera estimada: {prediccion_clas[0]}")


🏨 Ocupación hotelera estimada: Baja


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


**Nuevo enfoque**

### 📅 ¿Por qué es importante incluir la variable `Fecha` en nuestros modelos?

En nuestro análisis anterior no tuvimos en cuenta la columna `Fecha`, que contiene información como `"1982-12"`, representando el **año** y el **mes** de cada registro. Sin embargo, esta variable es muy importante porque:

- 📈 El **año** puede reflejar tendencias a largo plazo: por ejemplo, el turismo ha aumentado en las últimas décadas.
- 📆 El **mes** puede capturar **patrones estacionales**: Cartagena recibe más turistas en diciembre, junio o Semana Santa.

---

### 🔍 ¿Qué pasa si ignoramos la fecha?

Si eliminamos o ignoramos esta columna, el modelo:
- ❌ No entiende que hay **una evolución en el tiempo** (no distingue 1982 de 2022).
- ❌ No puede aprender **cuáles meses del año tienen más turismo**.
- ❌ Pierde contexto que podría mejorar su capacidad predictiva.

---

### 🛠️ ¿Qué haremos ahora?

Vamos a **extraer el año y el mes** de la columna `Fecha` para crear dos nuevas columnas:

- `Año`: Para detectar si el turismo ha crecido o disminuido a lo largo del tiempo.
- `Mes`: Para identificar patrones repetitivos como temporadas altas o bajas.

```python
# Convertir 'Fecha' a tipo datetime y crear nuevas columnas
df['Fecha'] = pd.to_datetime(df['Fecha'], format='%Y-%m')
df['Año'] = df['Fecha'].dt.year
df['Mes'] = df['Fecha'].dt.month


### 🎯 ¿Cómo mejora esto nuestros modelos?

| **Tipo de Modelo** | **Beneficio al usar Fecha** |
|--------------------|------------------------------|
| 🔢 **Regresión**     | Detecta tendencias (más turistas en años recientes, más en diciembre). |
| 🏷️ **Clasificación** | Aprende en qué meses hay más probabilidad de *Alta* o *Baja* ocupación. |

---

💡 **Incluir variables temporales es fundamental en modelos que analizan comportamientos a lo largo del tiempo**, como el turismo.

